File: Andrej/src/main/scala/nl/woupiestek/andrej/Free.scala
package nl.woupiestek.andrej

import nl.woupiestek.andrej.Free.Bind

import scala.annotation.tailrec
import scala.language.higherKinds

sealed trait Free[F[_], T] {
  def fold[U](x: Free.Folder[F, U], y: T => U): U

  final def flatMap[U](f: T => Free[F, U]): Free[F, U] =
    fold(new Free.Folder[F, Free[F, U]] {
      override def fold[W]
          : (F[W], (W) => Free[F, U]) => Bind[F, W, U] =
        Free.Bind[F, W, U]
    }, f)

  final def map[U](f: T => U): Free[F, U] =
    flatMap(x => Free(f(x)))
}

object Free {

  trait Folder[F[_], V] {
    def fold[U]: (F[U], U => V) => V
  }

  case class Return[F[_], T](value: T) extends Free[F, T] {
    override def fold[U](x: Folder[F, U], y: T => U): U =
      y(value)
  }

  case class Bind[F[_], E, T](
      request: F[E],
      callback: E => Free[F, T]
  ) extends Free[F, T] {
    override def fold[U](f: Folder[F, U], y: T => U): U =
      f.fold(request, (e: E) => callback(e).fold(f, y))
  }

  def apply[F[_], T](t: T): Free[F, T] = Return(t)

  def lift[F[_], T](ft: F[T]): Free[F, T] =
    Bind[F, T, T](ft, apply)
}

trait StraightInterpreter[F[_]] {
  def evaluate[T](ft: F[T]): T

  @tailrec final def execute[T](fft: Free[F, T]): T =
    fft match {
      case Free.Return(t) => t
      case Free.Bind(r, c) =>
        execute(c.asInstanceOf[Any => Free[F, T]](evaluate(r)))
    }
}

trait ReproducingInterpreter[F[_]] {

  def evaluate[T](ft: F[T]): (T, ReproducingInterpreter[F])

  @tailrec final def execute[T](fft: Free[F, T]): T =
    fft match {
      case Free.Return(t) => t
      case Free.Bind(r, c) =>
        evaluate(r) match {
          case (x, y) =>
            y.execute(c.asInstanceOf[Any => Free[F, T]](x))
        }
    }
}

trait CallbackInterpreter[F[_]] extends Free.Folder[F, Unit] {
  def fold[T](ft: F[T], cb: T => Unit): Unit

  //since this is not tail recursive, evaluate had better know what it is doing
  final def execute[T](fft: Free[F, T], cb: T => Unit): Unit =
    fft.fold(this, cb)
}

trait GoesInterpreter[F[_], G[_]] {

  def evaluate[T](t: F[T]): Free[G, T]

  def execute[T](fft: Free[F, T]): Free[G, T] = fft match {
    case Free.Return(t) => Free.Return(t)
    case Free.Bind(r, c) =>
      evaluate(r).flatMap(
        x => execute(c.asInstanceOf[Any => Free[F, T]](x))
      )
  }

}

File: Andrej/src/main/java/nl/woupiestek/andrej/java/Free.java
package nl.woupiestek.andrej.java;

import java.util.function.Consumer;
import java.util.function.Function;

public abstract class Free<T> {

    public final <U> Free<U> flatMap(Function<T, Free<U>> f) {
        return fold(new Folder<Free<U>>() {
            @Override
            public <Request, Response> Free<U> onRequest(Request request, Function<Response, Free<U>> function) {
                return new Bind<Request, Response, U>(request) {
                    @Override
                    public Free<U> getNext(Response s) {
                        return function.apply(s);
                    }
                };
            }
        }, f);
    }

    public final <U> Free<U> map(Function<T, U> f) {
        return flatMap(p -> new Unit<U>(f.apply(p)));
    }

    @FunctionalInterface
    public interface Eacher {
        <Request, Response> void onRequest(Request request, Consumer<Response> consumer);
    }

    public abstract void foreach(Eacher onBind, Consumer<T> onUnit);

    @FunctionalInterface
    public interface Folder<Target> {
        <Request, Response> Target onRequest(Request request, Function<Response, Target> function);
    }

    public abstract <U> U fold(Folder<U> onBind, Function<T, U> onUnit);

    public final boolean hasNext() {
        return fold(new Folder<Boolean>() {
            @Override
            public <Request, Response> Boolean onRequest(Request request, Function<Response, Boolean> function) {
                return true;
            }
        }, x -> false);
    }

    private static final class Unit<Value> extends Free<Value> {
        public final Value value;

        public Unit(Value value) {
            this.value = value;
        }

        @Override
        public void foreach(Eacher eacher, Consumer<Value> onUnit) {
            onUnit.accept(value);
        }

        @Override
        public <U> U fold(Folder<U> onBind, Function<Value, U> onUnit) {
            return onUnit.apply(value);
        }
    }

    private static abstract class Bind<Request, Response, Value> extends Free<Value> {
        public final Request request;

        Bind(Request request) {
            this.request = request;
        }

        public abstract Free<Value> getNext(Response s);

        @Override
        public final void foreach(Eacher onBind, Consumer<Value> onUnit) {
            onBind.onRequest(request, (Response x) -> getNext(x).foreach(onBind, onUnit));
        }

        @Override
        public final <U> U fold(Folder<U> onBind, Function<Value, U> onUnit) {
            return onBind.onRequest(request, (Response x) -> getNext(x).fold(onBind, onUnit));
        }
    }
}
